subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java

```diff
@@ -87,41 +98,61 @@ private String maybe(String prefix, String value) {
      *
      * @return a File object with the path to the archive
      */
-    @OutputFile
+    @Internal("Represented as a part of the archiveFile")
     public File getArchivePath() {
-        File destinationDir = getDestinationDir();
-        //noinspection ConstantConditions
-        if (destinationDir == null) {
-            throw new InvalidUserDataException("The destinationDir property must be set. Please apply the base plugin or set it explicitly.");
-        }
-        return new File(destinationDir, getArchiveName());
+        return getArchiveFile().get().getAsFile();
+    }
+
+    /**
+     * The {@link RegularFile} where the archive is constructed.
+     * The path is simply the {@code destinationDir} plus the {@code archiveName}.
+     *
+     * @implNote This returns a provider of {@link RegularFile} instead of {@link RegularFileProperty} in order to
+     * prevent users calling {@link org.gradle.api.provider.Property#set} and causing a plugin or users using
+     * {@link AbstractArchiveTask#getArchivePath()} to break or have strange behaviour.
+     * An example can be found <a href="https://github.com/gradle/gradle-native/issues/893#issuecomment-430776251">here</a>.
```

Comment:
I didn't really want this to show up in the generated docs, so I left it using the `implNote` tag, hope that's okay.